### PR TITLE
Add more details to testing bad practices

### PR DESCRIPTION
**Pull request by andrross**
These are a few cases I have seen that have resulted in flaky tests. I would love to see more details added here so that this can be used as a sort of checklist when writing, reviewing, or trying to fix tests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by reta — APPROVED** [review]

File: TESTING.md [hunk 1]
```diff
@@ -477,6 +477,18 @@ However, it should not be used for coverage. For instance if you are testing a p
 
 Multi-threaded tests are often not reproducible due to the fact that there is no guarantee on the order in which operations occur across threads. Adding randomization to the mix usually makes things worse and should be done with care.
 
+### Use `Thread.sleep`
+
+`Thread.sleep()` is almost always a bad idea because it is very difficult to know that you've waited long enough. Using primitives like `waitUntil` or `assertBusy`, which use Thread.sleep internally, is okay to wait for a specific condition. However, it is almost always better to instrument your code with concurrency primitives like a CountDownLatch that will allow you to deterministically wait for a specific condition, without waiting longer than necessary that will happen with a polling approach used by `assertBusy`.
```
Comment (dblock) [inline TESTING.md:482]: Quote `CountDownLatch`

File: TESTING.md [hunk 2]
```diff
@@ -477,6 +477,18 @@ However, it should not be used for coverage. For instance if you are testing a p
...
+
+### Expect a specific segment topology
+
+By design, OpenSearch integration tests will vary how the merge policy works because in almost all scenarios you should not depend on a specific segment topology (in the real world your code will see a huge diversity of indexing workloads with OpenSearch merging things in the background all the time!). If you do in fact need to care about the segment topology (e.g. for testing statistics that might vary slightly depending on number of segments), then you must take care to ensure that segment topology is deterministic by doing things like disabling background refreshes, force merging after indexing data, etc.
```
Comment (dblock) [inline TESTING.md:486]: Maybe take the text in parenthesis out into a sentence?

**Review by dblock — CHANGES_REQUESTED** [review]
Update TOC please?

It would be amazing if this advice came with links to good examples!

File: TESTING.md (hunk 2 above)
Comment (andrross) [reply TESTING.md:486]: Sure. I think I overuse parentheticals (maybe).

**Pull request merged**